### PR TITLE
Bug 1413149 - add beetmoverscript templates to support firefox releases

### PR DESCRIPTION
**Pull request by MihaiTabara**
+ tweaking it in puppet as well if looks alright.

### CODE REVIEW

**Comment by MihaiTabara** [comment]
For more context, out of stuff from [mozconfigs](https://hg.mozilla.org/mozilla-central/file/tip/testing/mozharness/configs/beetmover), we have the following:

- en_us_build.yml.tmpl
- en_us_repackage.yml.tmpl
- en_us_repackage_signing.yml.tmpl
- en_us_signing.yml.tmpl
- partials
- recompressed_completes.yml.tmpl

all reorganized to form `firefox_candidates.yml`.
And secondly, 

- recompressed_completes.yml.tmpl
- partials
- repacks

to form `firefox_candidates_repacks.yml`.

For some of the configs, I wasn't 100% sure what to fill-in so I left a bunch of TODOs and FIXME at which I'll come back later.

File: beetmoverscript/templates/firefox_candidates.yml [hunk 1]
```diff
@@ -0,0 +1,129 @@
+---
+metadata:
+    name: "Beet Mover Manifest"
+    description: "Maps {{ product }} artifacts to pretty names for en-US"
+    owner: "release@mozilla.com"
+
+#TODO: we might s/firefox/{product} + some code logic tweak to get devedition for free
```
Comment (escapewindow) [inline beetmoverscript/templates/firefox_candidates.yml]: Good idea. We don't have to do that now, but if devedition turns out to be close enough we can do that then.

File: beetmoverscript/templates/firefox_candidates.yml [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    target.crashreporter-symbols.zip:
+      s3_key: {{ platform }}/{{ locale }}/firefox-{{ version }}.crashreporter-symbols.zip
+      destinations:
+        - {{ platform }}/{{ locale }}/firefox-{{ version }}.crashreporter-symbols.zip
+    target.complete.mar:
+      s3_key: update/{{ platform }}/{{ locale }}/firefox-{{ version }}.complete.mar
+      destinations:
+        - update/{{ platform }}/{{ locale }}/firefox-{{ version }}.complete.mar
+      update_balrog_manifest: true
+    # FIXME: potentiall other names here
```
Comment (escapewindow) [inline beetmoverscript/templates/firefox_candidates.yml]: sp: potentially or potential

File: beetmoverscript/templates/firefox_candidates.yml [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    target.bz2.complete.mar:
+      s3_key: update/{{ platform }}/{{ locale }}/firefox-{{ version }}.bz2.complete.mar
+      destinations:
+        - update/{{ platform }}/{{ locale }}/firefox-{{ version }}.bz2.complete.mar
+    # FIXME: potentiall other names here
```
Comment (escapewindow) [inline beetmoverscript/templates/firefox_candidates.yml]: here too :)

File: beetmoverscript/templates/firefox_candidates_repacks.yml [hunk 4]
```diff
@@ -0,0 +1,64 @@
+---
+metadata:
+    name: "Beet Mover Manifest"
+    description: "Maps {{ product }} artifacts to pretty names for {{ locale }} L10n repack"
+    owner: "release@mozilla.com"
+
+#TODO: we might s/firefox/{product} + some code logic tweak to get devedition for free
```
Comment (escapewindow) [inline beetmoverscript/templates/firefox_candidates_repacks.yml]: Same comments in this file as the other one.

**Comment by MihaiTabara** [comment]
I've pushed another commit to address the comments, thanks!

I'll hold on for now as I don't have any in-tree tasks added. Once I have en-US completes beetmover jobs added at least or anything alike and I can start triggering some `maple`-based Firefox releases on staging, I'll go ahead and use one of your suggestions & pin the environment and start testing.

Thanks!

**Comment by MihaiTabara** [comment]
@escapewindow  With more than a month delay, I'm finally taking your advice and pushing this forward to merge to master. There's too many changes now and I want to make sure I set a backward compatible milestone before I move forward with push-to-cdns and whatever is next for cleanup.

I added small other bugfixes which would have broken the firefox nightlies. I think this version is good though. If you agree, I'll land in puppet and to production but keeping releaseduty in the loop with potential issues that may come up in Fennec beta/release and nightlies.

File: beetmoverscript/script.py [hunk 5]
```diff
@@ -292,6 +292,14 @@ def generate_balrog_info(context, artifact_pretty_name, locale, destinations, fr
     }
     if from_buildid:
         data["from_buildid"] = from_buildid
+        if is_promotion_action(context.action):
+            partials = get_partials_props(context.task)
+            # TODO: improve search by a dedicated dict with buildid as key
+            for p in partials.values():
+                if p['buildid'] == str(from_buildid):
+                    data['previousVersion'] = p['previousVersion']
+                    data['previousBuildNumber'] = p['previousBuildNumber']
```
Comment (escapewindow) [inline beetmoverscript/script.py]: Do we want a `break` here?

Also, do we have anything validating the `get_partials_props` schema? That may belong in `get_partials_props`, or in the release task schema as an optional field. This doesn't block, but may help in the future.

File: beetmoverscript/templates/firefox_candidates_repacks.yml [hunk 6]
```diff
@@ -0,0 +1,82 @@
+---
+metadata:
+    name: "Beet Mover Manifest"
+    description: "Maps {{ product }} artifacts to pretty names for {{ locale }} L10n repack"
+    owner: "release@mozilla.com"
+
+# TODO: once firefox is done, we might want to s/firefox/{product} + some code change to get devedition for free
```
Comment (escapewindow) [inline beetmoverscript/templates/firefox_candidates_repacks.yml]: I was just about to say this :)

**Comment by MihaiTabara** [comment]
I'll land as is and will address your comments in the next branch I'm already prepping for devedition/push-to-cdns and other stuff which will likely land by EOW.

**Pull request merged**